sdks/python/apache_beam/runners/dataflow/dataflow_metrics.py

```diff
@@ -30,25 +30,57 @@
 from apache_beam.metrics.metricbase import MetricName
 
 
-# TODO(pabloem)(JIRA-1381) Implement this once metrics are queriable from
-# dataflow service
 class DataflowMetrics(MetricResults):
   """Implementation of MetricResults class for the Dataflow runner."""
 
-  def __init__(self, dataflow_client=None, job_result=None):
+  def __init__(self, dataflow_client=None, job_result=None, job_graph=None):
     """Initialize the Dataflow metrics object.
 
     Args:
       dataflow_client: apiclient.DataflowApplicationClient to interact with the
         dataflow service.
       job_result: DataflowPipelineResult with the state and id information of
-        the job
+        the job.
+      job_graph: apiclient.Job instance to be able to translate between internal
+        step names (e.g. "s2"), and user step names (e.g. "split").
     """
     super(DataflowMetrics, self).__init__()
     self._dataflow_client = dataflow_client
     self.job_result = job_result
     self._queried_after_termination = False
     self._cached_metrics = None
+    self._job_graph = job_graph
+
+  def _translate_step_name(self, internal_name):
+    """Translate between internal step names (e.g. "s1") and user step names."""
+    if not self._job_graph:
+      raise ValueError('Could not translate the internal step name.')
+
+    try:
+      [step] = [step
+                for step in self._job_graph.proto.steps
+                if step.name == internal_name]
+      [user_step_name] = [prop.value.string_value
+                          for prop in step.properties.additionalProperties
+                          if prop.key == 'user_name']
+    except ValueError:
+      raise ValueError('Could not translate the internal step name.')
+    return user_step_name
+
+  def _get_metric_key(self, metric):
+    """Populate the MetricKey object for a queried metric result."""
+    try:
+      [step] = [prop.value
+                for prop in metric.name.context.additionalProperties
+                if prop.key == 'step']
+      step = self._translate_step_name(step)
+      [namespace] = [prop.value
+                     for prop in metric.name.context.additionalProperties
+                     if prop.key == 'namespace']
+      name = metric.name.name
+    except ValueError:
+      [step, namespace, name] = metric.name.name.split('/')
```

Comment:
Is there a way to make this more robust?  My belief is that step names may also contain '/' characters, in which case this raises a weird error.